helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -102,4 +107,36 @@ public boolean isShutdown() {
   public boolean isTerminated() {
     return _taskExecutor.isTerminated();
   }
+
+  /*
+   * Get target thread pool size from InstanceConfig first; if that fails, get it from
+   * ClusterConfig; if that fails, fall back to the default value.
+   */
+  private static int getTaskThreadPoolSize(HelixManager manager) {
+    ConfigAccessor configAccessor = manager.getConfigAccessor();
+    // Check instance config first for thread pool size
+    InstanceConfig instanceConfig =
+        configAccessor.getInstanceConfig(manager.getClusterName(), manager.getInstanceName());
+    if (instanceConfig != null) {
```

Comment:
nit, shall we throw exception if the instance config or cluster config is not found?
This mostly because the read is not done successfully. In which case, we don't know the real configured numbers. So better stop initialization instead of guess it shall be 40.

Reply:
For `InstanceConfig` it's actually valid to get null. When an instance is set up for the first time it's not going to have `InstanceConfig` when this code is executed, so an exception is not very appropriate because that situation is actually correct. 

I do think the `ClusterConfig` should have been there, but I still think an exception may be too strong.

Reply:
I am okay with returning the default size because creating a thread pool with a default size won't be the end of the world and won't affect the user too much. And the user can always try to set it again or restart if they want to change the thread pool size. 

But we should make it clear to the user. I don't see any log or notification here, so that could be a problem.

Reply:
So the context has changed, and what I said earlier is no longer valid. Please take a look if you want to further this discussion:

1. Thread pool creation is now after HelixManager connection, which means `InstanceConfig` and `ClusterConfig` are expected to exist. It should be treated as an error if they don't exist. 
2. I verified with test cases and believe that an earlier comment made by @narendly is incorrect: if `InstanceConfig` or `ClusterConfig` doesn't exist, `getInstanceConfig()`/`getClusterConfig()` will throw an HelixException because `isInstanceSetup()`/`isClusterSetup()` will fail. 

Here are the implications of these two points:

With the current design as shown here, if `InstanceConfig`/`ClusterConfig` doesn't exist, the attempt to create a thread pool will result an HelixException. I believe it's fine to **not** catch the exception and let it propagate up, because if `InstanceConfig`/`ClusterConfig` doesn't exist, it's an erroneous state. This aligns with what @jiajunwang is saying, and is the current design as we see it. 

Alternatively, we can catch the HelixException, log it, and use the default value (this is @narendly 's idea). 

I'm leaning towards the current design (let HelixException propagate upwards if configs don't exist). Anyone has a different opinion?

Reply:
Additionally, we are also updating `LiveInstance`, which I'm giving the same treatment. Read: if `InstanceConfig`, `ClusterConfig`, `LiveInstance` don't exist, the `TaskStateModelFactory` will **not** handle the exception and will let the process fail, because they should exist. I believe if they don't exist, somewhere else will also fail on participant side.

I'm open to your suggestions, what do you think? @jiajunwang @narendly